### PR TITLE
add script to embed basic ELF note metadata

### PR DESCRIPTION
**Pull request by anphel31**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/tools/cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
Embed Mariner ELF binaries with build metadata in a custom note section.
This initial version adds only the following metadata:
{
 "os": "mariner",
 "osVersion": "1.0",
 "type": "rpm"
}
For more information see: https://systemd.io/COREDUMP_PACKAGE_METADATA/

###### Change Log  

- Add gen-ld-script.sh to mariner-rpm-macros, calling the script before every SPEC build's prep section
- Change LDFLAGS to include the linker script created by gen-ld-script.sh

###### Does this affect the toolchain?  

**YES**

###### Test Methodology

- Local package build of some SPEC files such as zlib.
1) Build zlib
2) Unpack zlib RPM files
rpm2cpio zlib-1.2.11-3.cm1.x86_64.rpm | cpio -idmv
3) Read the note
$ readelf -n ./usr/lib/libz.so.1.2.11
Displaying notes found in: .note.package
  Owner                 Data size       Description
  FDO                  0x0000002c       Unknown note type: (0xcafe1a7e)
   description data: 7b 0a 20 22 6f 73 22 3a 20 22 43 42 4c 2d 4d 61 72 69 6e 65 72 22 2c 0a 20 22 76 65 72 73 69 6f 6e 22 3a 20 22 31 2e 30 22 0a 7d 00

### CODE REVIEW

File: SPECS/mariner-rpm-macros/macros [hunk 1]
```diff
@@ -32,6 +32,7 @@
 # are consistently applied this override should be removed.
 %__spec_build_template	#!%{__spec_build_shell}\
 %{__spec_build_pre}\
+/usr/lib/rpm/mariner/gen-ld-script.sh\
```
Comment (dmcilvaney) [inline SPECS/mariner-rpm-macros/macros:35]: is `%{mariner_release_version}` populated here? Possible enhancement might be to plumb that into the script and dynamically populate.

I've got some code somewhere that does hex translation in bash, I'll dig it up.

Comment (dmcilvaney) [reply SPECS/mariner-rpm-macros/macros:35]: Found it, was almost perfectly portable. Modified it to fit this use quickly.

```
#!/bin/bash

version=$1
[[ $version =~ ^[0-9]\.[0-9]$ ]] || { echo "$version is not a supported version"; exit 1; }

version_major=${version%%.*}
version_minor=${version##*.}

#Output digit without newline, od without address info (-An), and of type single byte hex (-t x1)
hex_major="0x$(printf $version_major | od -An -t x1 | xargs)"
hex_minor="0x$(printf $version_minor | od -An -t x1 | xargs)"

echo "BYTE($hex_major) BYTE(0x2E) BYTE($hex_minor)"
```

```
./release_test.sh 1.0
BYTE(0x31) BYTE(0x2E) BYTE(0x30)

./release_test.sh 134.456
123.456 is not a supported version (exit 1)
```

Comment (anphel31) [reply SPECS/mariner-rpm-macros/macros:35]: Thanks for sharing. Yes, ${mariner_release_version} is available here and I was able to pass it to the script. However, I plan to add these enhancements later on. Adding additional parameters with potentially different lengths means that the script will need to calculate the description length, add padding to get 4 byte alignment, etc.

Comment (dmcilvaney) [reply SPECS/mariner-rpm-macros/macros:35]: Agreed, its non-trivial to handle (I just barfed with exit 1 if it didnt fit the form 'x.y' exactly, with one digit for each

**Review by dmcilvaney — APPROVED** [review]

File: SPECS/mariner-rpm-macros/gen-ld-script.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# Embed the following metadata:
+#
+# {
+#  "os": "mariner",
+#  "osVersion": "1.0",
+#  "type": "rpm"
+# }
+
+LINKER_SCRIPT_START="SECTIONS { .note.package ALIGN(4): { "
+LINKER_SCRIPT_NAMESZ="BYTE(0x04) BYTE(0x00) BYTE(0x00) BYTE(0x00) "
```
Comment (jslobodzian) [inline SPECS/mariner-rpm-macros/gen-ld-script.sh:14]: Can we get additional comments around each of these....so we can see what the descriptions actually say/embed?

**Review by jslobodzian — APPROVED** [review]
approved with minor commenting change

File: SPECS/mariner-rpm-macros/gen-ld-script.sh (hunk 2 above)
Comment (anphel31) [reply SPECS/mariner-rpm-macros/gen-ld-script.sh:14]: Sure, I've added comments

**Pull request merged**